server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,296 @@
+            BlobStoreIndexShardSnapshot.FileInfo fileInfo = new BlobStoreIndexShardSnapshot.FileInfo(
+                name,
+                new StoreFileMetadata(name, uploadedSegmentMetadata.getLength(), uploadedSegmentMetadata.getChecksum(), Version.LATEST),
+                null
+            );
+            return new OnDemandBlockSnapshotIndexInput(fileInfo, localDirectory, transferManager);
+        }
+    }
+
+    /**
+     * Closes the directory
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void close() throws IOException {
+        Arrays.stream(localDirectory.listAll()).forEach(f -> fileCache.remove(localDirectory.getDirectory().resolve(f)));
+        localDirectory.close();
+        remoteDirectory.close();
+    }
+
+    /**
+     * Function to perform operations once files have been uploaded to Remote Store
+     * Currently deleting the local files here, as once uploaded to Remote, local files become eligible for eviction from FileCache
+     * @param files : recent files which have been successfully uploaded to Remote Store
+     * @throws IOException in case of I/O error
+     */
+    public void afterSyncToRemote(Collection<String> files) throws IOException {
+        logger.trace("afterSyncToRemote called for {}", files);
+        if (remoteDirectory == null) {
+            logger.trace("afterSyncToRemote called even though remote directory is not set");
+            return;
+        }
+        for (String fileName : files) {
+            /*
+            Decrementing the refCount here for the path so that it becomes eligible for eviction
+            This is a temporary solution until pinning support is added
+            TODO - Unpin the files here from FileCache so that they become eligible for eviction, once pinning/unpinning support is added in FileCache
+            Uncomment the below commented line(to remove the file from cache once uploaded) to test block based functionality
+             */
+            logger.trace("File {} uploaded to Remote Store and now can be eligible for eviction in FileCache", fileName);
+            fileCache.decRef(localDirectory.getDirectory().resolve(fileName));
+            // fileCache.remove(localDirectory.getDirectory().resolve(fileName));
+        }
+    }
+
+    private boolean isTempFile(String name) {
+        return name.endsWith(".tmp");
+    }
+
+    /**
+     * Return the list of files present in Remote
+     */
+    private String[] getRemoteFiles() throws IOException {
+        String[] remoteFiles;
+        try {
+            remoteFiles = remoteDirectory.listAll();
+        } catch (NullPointerException e) {
+            /*
+            There are two scenarios where the listAll() call on remote directory returns NullPointerException:
+            - When remote directory is not set
+            - When init() of remote directory has not yet been called
```

Comment:
We should not handle this NPE here, looks to be some bug, so trying to better understand this.
1. `CompositeDirectory` takes `RemoteDirectory` as input so it should never be null. We should add non-null check in the constructor to catch it early on. 
2. Why a non-initialized `RemoteDirectory` will be passed to the `CompositeDirectory`

Reply:
Maybe I should correct the explanation a bit here - the NPE occurs because the the [listAll](https://github.com/opensearch-project/OpenSearch/blob/c328c18f69c08b21aa8a76af270b04a70c5a8069/server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java#L376) implemented in the RemoteSegmentStoreDirectory is a bit buggy - if the latest metadata file read is empty (which can occur when no data has been uploaded to remote segment store yet), then it tries to extract info from the null metadata resulting in NPE.

This needs to be corrected in the listAll implementation of RemoteSegmentsStoreDirectory itself, will create a tracking task to correct it after exploring whether this NPE exception is being used to catch something in the remote directory flow or not.

Corrected the code comments for better explanation. Also the non null check is being done before initializing the Composite Directory in IndexService itself.

Reply:
> This needs to be corrected in the listAll implementation of RemoteSegmentsStoreDirectory itself, will create a tracking task to correct it after exploring whether this NPE exception is being used to catch something in the remote directory flow or not.

Yes we should fix this in RemoteSegmentsStoreDirectory instead of catching NPE here.